infra/src/Pos/Infra/InjectFail.hs

```diff
@@ -0,0 +1,84 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE OverloadedStrings          #-}
+-- | Fault injection, strategically placed at the root of CSL dependency graph.
+
+module Pos.Infra.InjectFail
+       ( FInjects
+       , parseFInjectsSpec
+       , mkFInjects
+       , FInject(..)
+       , testFInject
+       , setFInject
+       , logFInject
+       , testLogFInject
+       ) where
+
+import           Universum
+
+import           Data.Aeson (FromJSON, ToJSON)
+import           Data.Char (toLower)
+import           Data.Maybe (catMaybes)
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import qualified Data.Text as T
+import           Options.Applicative
+
+import           Pos.Util.Wlog (CanLog, HasLoggerName, logError,
+                     modifyLoggerName)
+
+
+
+data FInject
+  = FInjIgnoreAPI                     -- ^ Return a hard-coded string for all registered Wallet API endpoints
+  | FInjIgnoreShutdown                -- ^ Ignore the shutdown request
+  | FInjApplyUpdateNoExit             -- ^ Don't exit after handling the 'update/apply' endpoint
+  | FInjApplyUpdateWrongExitCode      -- ^ Exit with a wrong exit code as response to the 'update/apply' request
+  deriving (Bounded, Enum, Eq, Generic, Ord, Show)
+instance FromJSON FInject
+instance ToJSON   FInject
+
+desc :: FInject -> String
+desc = \case
+    FInjIgnoreAPI                     ->  "Return a hard-coded string for all registered Wallet API endpoints"
+    FInjIgnoreShutdown                ->  "Ignore the shutdown request"
+    FInjApplyUpdateNoExit             ->  "Don't exit after handling the 'update/apply' endpoint"
+    FInjApplyUpdateWrongExitCode      ->  "Exit with a wrong exit code as response to the 'update/apply' request"
+
+newtype FInjects = FInjects { _fromFInjs :: IORef (Set FInject) }
```

Comment:
I'd be really inclined to seeing that this is intrinsically full of side-effects. What about encapsulating the API in a data-type with a target monad as parameter:


```hs
newtype FInjects m = FInjects 
    { testFInject    :: FInject -> m Bool
    , setFInject     :: FInject -> m ()
    , logFInject     :: FInject -> m ()
    , testLogFInject :: FInject -> m Bool
    }
```

With, `mkInjects` defined as:

```hs
-- | Make a stateful fault injection configuration object.
mkFInjects :: MonadIO m => Set FInject -> m (FInjects m)
mkFInjects fs = do
    ioRef <- liftIO $ newIORef fs
    return $ FInjects
        { testFInject = mkTestFInject ioRef
        , -- ... and so forth 
        }
```

This way, we can easily mock this if needed and, above all, we clearly mark it as something that needs a given context to run, which is obvious at first sight:

```hs
data ShutdownContext = ShutdownContext
    { _shdnIsTriggered :: !(TVar Bool)
    , _shdnFInjects    :: !FInjects IO
    -- ^ If this flag is `True`, then workers should stop.
    }
```

Reply:
Done!